sdk/storage/Azure.Storage.Internal.Avro/src/AvroReader.cs

```diff
@@ -0,0 +1,196 @@
+
+        /// <summary>
+        /// Avro schema.
+        /// </summary>
+        private AvroType _itemType;
+
+        /// <summary>
+        /// The number of items remaining in the current block.
+        /// </summary>
+        private long _itemsRemainingInBlock;
+
+        /// <summary>
+        /// The byte offset within the Avro file (both header and data)
+        /// of the start of the current block.
+        /// </summary>
+        public long BlockOffset { get; private set; }
+
+        /// <summary>
+        /// The index of the current object within the current block.
+        /// </summary>
+        /// <returns></returns>
+        public long ObjectIndex { get; private set; }
+
+        /// <summary>
+        /// If this Avro Reader has been initalized.
+        /// </summary>
+        private bool _initalized;
+
+        /// <summary>
+        /// Constructor for an AvroReader that will read from the
+        /// beginning of an Avro file.
+        /// </summary>
+        public AvroReader(Stream dataStream)
+        {
+            _dataStream = dataStream;
+            _headerStream = dataStream;
+            _metadata = new Dictionary<string, string>();
+            _initalized = false;
+        }
+
+        /// <summary>
+        /// Constructor for an Avro Reader that will read beginning
+        /// in the middle of an Avro file.
+        /// </summary>
+        public AvroReader(
+            Stream dataStream,
+            Stream headerStream,
+            long currentBlockOffset,
+            long indexWithinCurrentBlock)
+        {
+            _dataStream = dataStream;
+            _headerStream = headerStream;
+            _metadata = new Dictionary<string, string>();
+            _initalized = false;
+            BlockOffset = currentBlockOffset;
+            ObjectIndex = indexWithinCurrentBlock;
+            _initalized = false;
+        }
+
+        private async Task Initalize(bool async, CancellationToken cancellationToken = default)
+        {
+            // Four bytes, ASCII 'O', 'b', 'j', followed by 1.
+            byte[] header = await AvroParser.ReadFixedBytesAsync(_headerStream, AvroConstants.InitBytes.Length, async, cancellationToken).ConfigureAwait(false);
+            if (!header.SequenceEqual(AvroConstants.InitBytes))
+            {
+                throw new ArgumentException("Stream is not an Avro file.");
+            }
+
+            // File metadata is written as if defined by the following map schema:
+            // { "type": "map", "values": "bytes"}
+            _metadata = await AvroParser.ReadMapAsync(_headerStream, AvroParser.ReadStringAsync, async, cancellationToken).ConfigureAwait(false);
+
+            // Validate codec
+            _metadata.TryGetValue(AvroConstants.CodecKey, out string codec);
+            if (codec == AvroConstants.DeflateCodec)
+            {
+                throw new ArgumentException("Deflate codec is not supported");
+            }
```

Comment:
I'd rather inverse this validation as there are more codecs available. 
See https://github.com/apache/avro/blob/6b0d470a79b4c4e10d9890183d6c913608a2a225/lang/py3/avro/datafile.py#L94-L99

Reply:
Good catch, fixed.

Reply:
Fixed.